python/pyspark/shuffle.py

```diff
@@ -367,32 +372,13 @@ def iteritems(self):
 
     def _external_items(self):
         """ Return all partitioned items as iterator """
-        assert not self.data
```

Comment:
I noticed that you moved a few of these assertions.  I guess the old assumption was that once we've spilled, we'll stop using `data` and only aggregate into `pdata`, given that we clear `data` in the first branch of `_spill`.  Why has this assumption changed here?  It looks like we do end up writing to `data` again inside of this `_external_items` method, but then we end up clearing `data` at the end after the iterator has been consumed.

Was this change necessary in order to support iterating multiple times over the merged result?  Just want to double-check my understanding here.